### PR TITLE
#66 Implemented methods in Field and NonogramSolver

### PR DESCRIPTION
**Pull request by MaximSmolskiy**

### CODE REVIEW

File: app/src/main/java/com/mithridat/nonoconverter/backend/Field.java [hunk 1]
```diff
@@ -138,8 +138,18 @@ public void setColor(int i, int j, int col) {
      *               COL, if column
      * @return index of first cell with another color in row or column
      */
-    public int getAntoherColorIndex(int ind, int pos, int dir, int type) {
-        return 0;
+    public int getAnotherColorIndex(int ind, int pos, int dir, int type) {
+        int i = pos + 1;
+        if (type == ROW) {
+            for(; i >= 0 && i < _cols; i += dir) {
```
Comment (maik-nack) [inline app/src/main/java/com/mithridat/nonoconverter/backend/Field.java]: Where is space? Check all your loops.

File: app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java [hunk 2]
```diff
@@ -75,17 +75,58 @@ public void setAsyncTask(AsyncTask<Void, Void, Field> asyncTask) {
...
     /**
      * Method for initialising class fields before solution start
      */
     void init() {
-
+        int rows = _nono.getLeftRowsLength(), cols = _nono.getTopColsLength();
+        int length;
+        _field = new Field(rows, cols);
+        _left = new boolean[rows][];
+        for(int i = 0; i < rows; i++) {
```
Comment (maik-nack) [inline app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java]: Is it possible to create one function for this and next loops?

File: app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java [hunk 3]
```diff
@@ -75,17 +75,58 @@ public void setAsyncTask(AsyncTask<Void, Void, Field> asyncTask) {
...
+        _top = new boolean[cols][];
+        for(int i = 0; i < cols; i++) {
+            length = _nono.getTopColLength(i);
+            _top[i] = new boolean[length];
+            for(int j = 0; j < length; j++) {
+                _top[i][j] = false;
+            }
+        }
+        _rows = new HashSet<>();
+        for(int i = 0; i < rows; i++) {
```
Comment (maik-nack) [inline app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java]: Is it possible to create one function for this and next loops?

File: app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java [hunk 4]
```diff
@@ -75,17 +75,58 @@ public void setAsyncTask(AsyncTask<Void, Void, Field> asyncTask) {
...
+            for(int j = 0; j < _left[i].length; j++) {
+                if(!_left[i][j]) return false;
+            }
+        }
+        for(int i = 0; i < _top.length; i++) {
+            for(int j = 0; j < _top[i].length; j++) {
+                if(!_top[i][j]) return false;
+            }
+        }
+        return true;
```
Comment (maik-nack) [inline app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java]: There is a calling of `check` method must be here.

File: app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java [hunk 5]
```diff
@@ -75,17 +75,58 @@ public void setAsyncTask(AsyncTask<Void, Void, Field> asyncTask) {
...
+        boolean isChanged = true;
+        init();
+        while(!_asyncTask.isCancelled() && isChanged) {
+            isChanged = applyAlgorithms(_rows, Field.ROW);
+            _rows.clear();
+            isChanged = applyAlgorithms(_cols, Field.COL) || isChanged;
+            _cols.clear();
         }
-        return false;
+        for(int i = 0; i < _left.length; i++) {
```
Comment (maik-nack) [inline app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java]: Is it possible to move this and next loops in `check` method?

File: app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java [hunk 6]
```diff
@@ -218,4 +259,32 @@ boolean applyDualPosition(int ind, int type) {
...
+     * @param set - set of indexes
+     * @param type - ROW, if row
+     *      *        COL, if column
+     * @return true, if any cell was filled
+     *         false, otherwise
+     */
+    private boolean applyAlgorithms(HashSet<Integer> set, int type) {
+        boolean isFilled = false, isChanged;
+        for(int i : set) {
+            isChanged = true;
```
Comment (maik-nack) [inline app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java:251]: Is it possible to check here cancelling of async task?

File: app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java [hunk 7]
```diff
@@ -218,4 +259,32 @@ boolean applyDualPosition(int ind, int type) {
         return false;
     }
 
+    /**
+     * For every row or column with index in set applies algorithms
+     *
+     * @param set - set of indexes
+     * @param type - ROW, if row
+     *      *        COL, if column
```
Comment (maik-nack) [inline app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java]: Two asterisks?

File: app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java [hunk 8]
```diff
@@ -75,17 +75,32 @@ public void setAsyncTask(AsyncTask<Void, Void, Field> asyncTask) {
...
+        return check();
     }
 
     /**
      * Method for initialising class fields before solution start
      */
     void init() {
-
+        int rows = _nono.getLeftRowsLength(), cols = _nono.getTopColsLength();
+        int length;
```
Comment (maik-nack) [inline app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java]: It is unused variable

File: app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java [hunk 9]
```diff
@@ -95,7 +110,17 @@ void init() {
      *         false, otherwise
      */
     boolean check() {
-        return true;
+        for (int i = 0; i < _left.length; i++) {
```
Comment (maik-nack) [inline app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java]: Is it possible to create one function for this and next loops? You can use foreach in these loops.

File: app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java [hunk 10]
```diff
@@ -218,4 +243,65 @@ boolean applyDualPosition(int ind, int type) {
...
+     * @param set - set of indexes
+     * @param type - ROW, if row
+     *               COL, if column
+     * @return true, if any cell was filled
+     *         false, otherwise
+     */
+    private boolean applyAlgorithms(HashSet<Integer> set, int type) {
+        boolean isFilled = false, isChanged;
+        for (int i : set) {
+            if (_asyncTask.isCancelled()) {
```
Comment (maik-nack) [inline app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java]: It must be one line

File: app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java [hunk 11]
```diff
@@ -218,4 +243,65 @@ boolean applyDualPosition(int ind, int type) {
...
+     * @param type - ROW, if _left
+     *               COL, if _top
+     */
+    private void initBooleanArray(boolean[][] array, int type) {
+        int length;
+        for (int i = 0; i < array.length; i++) {
+            length = type == Field.ROW
+                    ? _nono.getLeftRowLength(i) : _nono.getTopColLength(i);
+            array[i] = new boolean[length];
+            for (int j = 0; j < length; j++) {
```
Comment (maik-nack) [inline app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java]: You can use `Arrays.fill`

File: app/src/main/java/com/mithridat/nonoconverter/backend/Field.java [hunk 12]
```diff
@@ -31,8 +31,8 @@
      */
     public Field(int rows, int cols) {
         _field = new int[rows][cols];
-        for(int i = 0; i < rows; i++) {
-            for(int j = 0; j < cols; j++) {
+        for (int i = 0; i < rows; i++) {
+            for (int j = 0; j < cols; j++) {
                 _field[i][j] = Colors.EMPTY;
```
Comment (maik-nack) [inline app/src/main/java/com/mithridat/nonoconverter/backend/Field.java]: You can use `Arrays.fill`

File: app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java [hunk 13]
```diff
@@ -218,4 +243,65 @@ boolean applyDualPosition(int ind, int type) {
...
+        }
+    }
+
+    /**
+     * Method for initialising one of the two hash sets
+     *
+     * @param set - hash set for initialising
+     * @param size - size of hash set
+     */
+    private void initHashSet(HashSet<Integer> set, int size) {
```
Comment (maik-nack) [inline app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java]: Is it possible to change this method to method for adding in set numbers from the interval?

File: app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java [hunk 14]
```diff
@@ -287,21 +277,38 @@ private void initBooleanArray(boolean[][] array, int type) {
...
+     * with numbers from interval
      *
      * @param set - hash set for initialising
-     * @param size - size of hash set
+     * @param begin - first number from interval
+     * @param end - next to last number from interval
      */
-    private void initHashSet(HashSet<Integer> set, int size) {
-        for (int i = 0; i < size; i++) {
+    private void initHashSet(HashSet<Integer> set, int begin, int end) {
```
Comment (maik-nack) [inline app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java]: But it is not `init`

**Review by maik-nack — APPROVED** [review]
It is a very good job!

**Comment by maik-nack** [comment]
In the future, we may add the following methods for convenience:
* in the `Field.java`
```java
    public int getLength(int type) {
        return type == Field.ROW ? getRows() : getCols();
    }

    public int getColor(int i, int j, int type) {
        return type == Field.ROW ? getColor(i, j) : getColor(j, i);
    }

    public void setColor(int i, int j, int col, int type) {
        if (type == Field.ROW) {
            setColor(i, j, col);
        } else {
            setColor(j, i, col);
        }
    }
```
* in the `Nonogram.java`
```java
    int getFirstLength(int type) {
        return type == Field.ROW ? getLeftRowsLength() : getTopColsLength();
    }

    int getSecondLength(int i, int type) {
        return type == Field.ROW ? getLeftRowLength(i) : getTopColLength(i);
    }

    int getValue(int i, int j, int type) {
        return type == Field.ROW ? getLeftRowValue(i, j) : getTopColValue(j, i);
    }
```

**Review by PawelMelnik — APPROVED** [review]
Nice

**Pull request merged**